service/history/workflow/mutable_state_impl.go

```diff
@@ -4236,6 +4236,52 @@ func (ms *MutableStateImpl) RejectWorkflowExecutionUpdate(_ string, _ *updatepb.
 	return nil
 }
 
+func (ms *MutableStateImpl) AddWorkflowExecutionOptionsUpdatedEvent(
+	versioningOverride *workflowpb.VersioningOverride,
+) (*historypb.HistoryEvent, error) {
+	if err := ms.checkMutability(tag.WorkflowActionWorkflowOptionsUpdated); err != nil {
+		return nil, err
+	}
+	event := ms.hBuilder.AddWorkflowExecutionOptionsUpdatedEvent(versioningOverride)
+	if err := ms.ApplyWorkflowExecutionOptionsUpdatedEvent(event); err != nil {
+		return nil, err
+	}
+	return event, nil
+}
+
+func (ms *MutableStateImpl) ApplyWorkflowExecutionOptionsUpdatedEvent(event *historypb.HistoryEvent) error {
+	override := event.GetWorkflowExecutionOptionsUpdatedEventAttributes().GetVersioningOverride()
+	previousCurrentDeployment := ms.GetCurrentDeployment()
+	if override != nil {
+		ms.GetExecutionInfo().VersioningInfo.BehaviorOverride = override.GetBehavior()
+		ms.GetExecutionInfo().VersioningInfo.DeploymentOverride = override.GetDeployment()
+	} else {
+		ms.GetExecutionInfo().VersioningInfo.BehaviorOverride = enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED
+		ms.GetExecutionInfo().VersioningInfo.DeploymentOverride = nil
+	}
+	// If behavior was changed, deployment will also have changed, so we only need to check deployment
+	if !proto.Equal(ms.GetCurrentDeployment(), previousCurrentDeployment) {
+		// In the case of deployment change, I'm choosing to let started WFTs that are running on the old deployment finish running
+		// The argument for failing them is if the user wants to reschedule them on the new deployment, but that is an optimization
+		// that can come later, especially since it's possible the user would prefer to not fail them.
+		// TODO (carly) part 2: if safe mode, do replay test on new deployment if deployment changed, if fail, revert changes and abort
+		ms.ClearStickyTaskQueue()
+		// TODO (carly): confirm this is the right way to reschedule pending WFT. Is there only one WFT?
+		if ms.HasPendingWorkflowTask() && !ms.HasStartedWorkflowTask() &&
+			// does not support them.
+			ms.GetPendingWorkflowTask().Type != enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE {
```

Comment:
should we reschedule pending speculative task as well? I don't the one scheduled on the old deployment will be able to start.

Maybe call GenerateScheduleSpeculativeWorkflowTaskTasks instead in case of speculative? cc @alexshtin

Reply:
changed the code, will make a comment on the new version taggin alex and asking him to approve the handling of speculative

Reply:
`GenerateScheduleSpeculativeWorkflowTaskTasks` only creates timeout task. It doesn't add WFT to matching. I think we can leave it as is (at least for now). Short (5s) delay might be there in worse case, but I think it is acceptable during deployments. Same thing can happen during matching deployment, btw.